Dask executor is broken with dask upstream

In our Dask executor, we create a high-level graph as follows:

https://github.com/pangeo-forge/pangeo-forge-recipes/blob/6194c72343d82e39b09edcafbe7085b1d9464787/pangeo_forge_recipes/executors/dask.py#L74-L76

Our upstream test (dask commit `725110f9367931291a3e68c9d582544cdb032f77`) has revealed the following [error](https://github.com/pangeo-forge/pangeo-forge-recipes/runs/4793899318?check_suite_focus=true), triggered by the line

```python
delayed = Delayed(prev_token[0], hlg)
```


```
self = Delayed('finalize_target-8dfa49acbe8cb65c6a2f82ebca37350d')
key = 'finalize_target-8dfa49acbe8cb65c6a2f82ebca37350d'
dsk = HighLevelGraph with 5 layers.
<dask.highlevelgraph.HighLevelGraph object at 0x7f338c0b3280>
 0. config
 1. cache_input
 2. prepare_target
 3. store_chunk
 4. finalize_target

length = None, layer = None

    def __init__(self, key, dsk, length=None, layer=None):
        self._key = key
        self._dask = dsk
        self._length = length
    
        # NOTE: Layer is used by `to_delayed` in other collections, but not in normal Delayed use
        self._layer = layer or key
        if isinstance(dsk, HighLevelGraph) and self._layer not in dsk.layers:
>           raise ValueError(
                f"Layer {self._layer} not in the HighLevelGraph's layers: {list(dsk.layers)}"
            )
E           ValueError: Layer finalize_target-8dfa49acbe8cb65c6a2f82ebca37350d not in the HighLevelGraph's layers: ['config', 'cache_input', 'prepare_target', 'store_chunk', 'finalize_target']
```

It looks like Dask does not like how we are creating the Delayed object. Is this a dask regression? Or do we need to change our code.

@TomAugspurger - any insights here?

When I try that with dask==2021.08.1, I get

```python
delayed = Delayed(prev_token[0], hlg, layer=stage.name)
E       TypeError: __init__() got an unexpected keyword argument 'layer'
```

--- a/pangeo_forge_recipes/executors/dask.py
+++ b/pangeo_forge_recipes/executors/dask.py
@@ -51,7 +51,6 @@ class DaskPipelineExecutor(PipelineExecutor[Delayed]):
         prev_dependency = ()  # type: Union[Tuple[()], Tuple[str]]
         for stage in pipeline.stages:
             stage_graph = {}  # type: Dict[Union[str, Tuple[str, int]], Any]
-            # each block should
             if stage.mappable is None:
                 stage_key = append_token(stage.name, token)
                 func = wrap_standalone_task(stage.function)

--- a/pangeo_forge_recipes/executors/dask.py
+++ b/pangeo_forge_recipes/executors/dask.py
@@ -43,37 +43,35 @@ class DaskPipelineExecutor(PipelineExecutor[Delayed]):
 
         # start with just the config as a standalone layer
         # create a custom delayed object for the config
-        config_task = "config"
-        config_token = append_token(config_task, token)
-        layers[config_task] = {config_token: pipeline.config}
-        dependencies[config_task] = set()
+        config_key = append_token("config", token)
+        layers[config_key] = {config_key: pipeline.config}
+        dependencies[config_key] = set()
 
-        prev_token = ()  # type: Tuple[str, ...]
-        prev_task = config_task
+        prev_key = config_key
+        prev_dependency = ()  # type: Union[Tuple, Tuple[str]]
         for stage in pipeline.stages:
-            stage_token = append_token(stage.name, token)
             stage_graph = {}  # type: Dict[Union[str, Tuple[str, int]], Any]
+            # each block should
             if stage.mappable is None:
+                stage_key = append_token(stage.name, token)
                 func = wrap_standalone_task(stage.function)
-                stage_graph[stage_token] = (func, config_token) + prev_token
-                prev_token = (stage_token,)
+                stage_graph[stage_key] = (func, config_key) + prev_dependency
             else:
                 func = wrap_map_task(stage.function)
                 checkpoint_args = []
                 for i, m in enumerate(stage.mappable):
-                    key = (stage.name, i)
-                    stage_graph[key] = (func, m, config_token) + prev_token
+                    key = (append_token(stage.name, token), i)
+                    stage_graph[key] = (func, m, config_key) + prev_dependency
                     checkpoint_args.append(key)
-                checkpoint_key = f"{stage.name}-checkpoint-{token}"
-                stage_graph[checkpoint_key] = (checkpoint, *checkpoint_args)
-                prev_token = (checkpoint_key,)
-            layer_name = prev_token[0]
-            layers[layer_name] = stage_graph
-            dependencies[layer_name] = {prev_task}
-            prev_task = stage.name
+                stage_key = f"{stage.name}-checkpoint-{token}"
+                stage_graph[stage_key] = (checkpoint, *checkpoint_args)
+            layers[stage_key] = stage_graph
+            dependencies[stage_key] = {config_key} | {prev_key}
+            prev_dependency = (stage_key,)
+            prev_key = stage_key
 
         hlg = HighLevelGraph(layers, dependencies)
-        delayed = Delayed(layer_name, hlg)
+        delayed = Delayed(prev_key, hlg)
         return delayed
 
     @staticmethod

--- a/pangeo_forge_recipes/executors/dask.py
+++ b/pangeo_forge_recipes/executors/dask.py
@@ -67,12 +67,13 @@ class DaskPipelineExecutor(PipelineExecutor[Delayed]):
                 checkpoint_key = f"{stage.name}-checkpoint-{token}"
                 stage_graph[checkpoint_key] = (checkpoint, *checkpoint_args)
                 prev_token = (checkpoint_key,)
-            layers[stage.name] = stage_graph
-            dependencies[stage.name] = {prev_task}
+            layer_name = prev_token[0]
+            layers[layer_name] = stage_graph
+            dependencies[layer_name] = {prev_task}
             prev_task = stage.name
 
         hlg = HighLevelGraph(layers, dependencies)
-        delayed = Delayed(prev_token[0], hlg)
+        delayed = Delayed(layer_name, hlg)
         return delayed
 
     @staticmethod

--- a/pangeo_forge_recipes/executors/dask.py
+++ b/pangeo_forge_recipes/executors/dask.py
@@ -48,7 +48,7 @@ class DaskPipelineExecutor(PipelineExecutor[Delayed]):
         dependencies[config_key] = set()
 
         prev_key = config_key
-        prev_dependency = ()  # type: Union[Tuple, Tuple[str]]
+        prev_dependency = ()  # type: Union[Tuple[()], Tuple[str]]
         for stage in pipeline.stages:
             stage_graph = {}  # type: Dict[Union[str, Tuple[str, int]], Any]
             # each block should

